### PR TITLE
API Server & kubectl logic for Vertical Pod Autoscaler

### PR DESCRIPTION
**Pull request by kgrygiel**
**What this PR does / why we need it**:
Adds API Server and kubectl logic to handle Vertical Pod Autoscalers.

**Special notes for your reviewer**:
This is follow up to PR [#63797](https://github.com/kubernetes/kubernetes/pull/63797) that adds VPA to the autoscaling v2beta1 API.
/cc @mwielgus 
/cc @wojtek-t 

**Release note**:
```
Support for Vertical Pod Autoscalers in the API Server and kubectl.
```

### CODE REVIEW

File: pkg/printers/internalversion/describe.go [hunk 1]
```diff
@@ -2962,6 +2963,117 @@ func describeHorizontalPodAutoscaler(hpa *autoscaling.HorizontalPodAutoscaler, e
...
 }
 
+// VerticalPodAutoscalerDescriber generates information about a vertical pod autoscaler.
+type VerticalPodAutoscalerDescriber struct {
+	client clientset.Interface
+}
+
+// Describe describes the vertical pod autoscaler object.
+func (d *VerticalPodAutoscalerDescriber) Describe(namespace, name string, describerSettings printers.DescriberSettings) (string, error) {
+	vpa, err := d.client.Autoscaling().VerticalPodAutoscalers(namespace).Get(name, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (kgrygiel) [reply pkg/printers/internalversion/describe.go]: Thanks! Changed to use the external client.

**Comment by mwielgus** [comment]
This has to be merged to not leave the autoscaling api in half-working/half-broken state.

**Comment by mwielgus** [comment]
Adding status/approved-for-milestone as a sig autoscaling leader.

**Comment by jberkus** [comment]
/hold

This is a 6K line patch you're delivering in the middle of code freeze.  When we talked about this, you said that this PR was just for tracking and wasn't going to merge.  Please discuss this on #sig-release on Slack.

**Comment by liggitt** [comment]
these two comments are in direct opposition:

https://github.com/kubernetes/kubernetes/pull/63797#issuecomment-395258077
> The risk associated with this PR is minimal. It is the api that is handled by an opt-in, independent component.

https://github.com/kubernetes/kubernetes/pull/64286#issuecomment-395497439:
> This has to be merged to not leave the autoscaling api in half-working/half-broken state.

if https://github.com/kubernetes/kubernetes/pull/63797 destabilized things and requires large follow-ups like this, we should revert it asap and bring it in in 1.12

File: pkg/printers/internalversion/describe.go [hunk 2]
```diff
@@ -2982,6 +2983,128 @@ func describeHorizontalPodAutoscaler(hpa *autoscaling.HorizontalPodAutoscaler, e
...
+	if describerSettings.ShowEvents {
+		events, _ = d.Core().Events(namespace).Search(legacyscheme.Scheme, vpa)
+	}
+
+	internalVpa := &autoscaling.VerticalPodAutoscaler{}
+	if err := legacyscheme.Scheme.Convert(vpa, internalVpa, nil); err != nil {
+		return "", err
+	}
+
+	return describeVerticalPodAutoscaler(internalVpa, events, d)
```
Comment (janetkuo) [inline pkg/printers/internalversion/describe.go]: You only use internal vpa. Therefore you don't need external client or conversion.

File: pkg/printers/internalversion/printers.go [hunk 3]
```diff
@@ -332,6 +332,15 @@ func AddHandlers(h printers.PrintHandler) {
 	h.TableHandler(horizontalPodAutoscalerColumnDefinitions, printHorizontalPodAutoscaler)
 	h.TableHandler(horizontalPodAutoscalerColumnDefinitions, printHorizontalPodAutoscalerList)
 
+	verticalPodAutoscalerColumnDefinitions := []metav1beta1.TableColumnDefinition{
+		{Name: "Name", Type: "string", Format: "name", Description: metav1.ObjectMeta{}.SwaggerDoc()["name"]},
+		{Name: "Selector", Type: "string", Description: autoscalingv2beta1.VerticalPodAutoscalerSpec{}.SwaggerDoc()["selector"]},
+		{Name: "UpdateMode", Type: "string", Description: autoscalingv2beta1.PodUpdatePolicy{}.SwaggerDoc()["updateMode"]},
```
Comment (janetkuo) [inline pkg/printers/internalversion/printers.go]: Selector, and even UpdateMode could be priority 1. `kubectl get` is useful when it prints the current status of a resource. Is there something in vpa status that you can add here? For example, deployment printer prints its desired and current replicas; pod printer prints its current ready containers and restart counts.

**Comment by mwielgus** [comment]
As discussed on slack we agreed to postpone it to 1.12. 

As of the comment of direct opposition - iVPA follows the same model as Ingress. The API itself is official, however various parties may provide slightly different implementation that fits to their needs. So it is both core and independent at the same time. Sorry if it wasn't crystal clear. 

The exact communication with #63797 was that there will be a follow up PR. @jberkus understood that the #63797 is that follow up PR. The miscommunication was resolved on slack.

**Pull request closed without merging**